### PR TITLE
MAINT: Simple speed-ups for getting overloaded types

### PR DESCRIPTION
**Pull request by mhvk**
Mostly aim to ensure that the most common case, in which the only override class is ndarray, is fast.

### CODE REVIEW

File: numpy/core/overrides.py [hunk 1]
```diff
@@ -45,12 +45,21 @@ def get_overloaded_types_and_args(relevant_args):
...
+                for i, old_arg in enumerate(overloaded_args):
+                    if issubclass(arg_type, type(old_arg)):
+                        index = i
+                        break
+                overloaded_args.insert(index, arg)
+            else:
+                overloaded_args = [arg]
+
+    # Short-cut for the common case of only ndarray.
+    if overloaded_types == [ndarray]:
```
Comment (shoyer) [inline numpy/core/overrides.py]: Should we save this list instead of recreating it?

**Comment by shoyer** [comment]
Nice! Do you see a measurable difference on our benchmarks for `__array_function__`?

File: numpy/core/overrides.py (hunk 1 above)
Comment (mhvk) [reply numpy/core/overrides.py]: Done.

**Comment by mhvk** [comment]
Timing for the hstack benchmark:
```
l = [np.arange(1000), np.arange(1000)]
%timeit np.hstack(l)
# 1.14.5: 100000 loops, best of 3: 2.79 µs per loop
# master: 100000 loops, best of 3: 9.96 µs per loop
# this PR: 100000 loops, best of 3: 7.43 µs per loop
```
So, an improvement but not a huge one.

**Review by shoyer — APPROVED** [review]

**Comment by shoyer** [comment]
@mhvk definitely a step in the right direction, thanks!

**Pull request merged**